File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int max_path = INT_MIN;
    
    int subtree(TreeNode * root) {
        if (root == NULL)
            return 0;
        int leftST = subtree(root->left);
        int rightST = subtree(root->right);
        max_path = max(max_path, leftST + rightST + root->val);
        return max({leftST + root->val, rightST + root->val, 0});
    }
    
    int maxPathSum(TreeNode* root) {
        subtree(root);
        return max_path;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        max_path = -float('inf')
        
        def subtree(node: Optional[TreeNode]) -> int:
            nonlocal max_path
            if not node:
                return 0
            
            leftST = max(subtree(node.left), 0)
            rightST = max(subtree(node.right), 0)
            
            max_path = max(max_path, leftST +rightST + node.val)
            
            return max(leftST + node.val, rightST + node.val)
        
        subtree(root)
        return max_path
            
        
        
        
        
